### PR TITLE
JPA Map Global Locking 

### LINKED ISSUE #14734: Introduce JPA locking mechanism implementation
Based on #9388, implement JPA locking mechanism

### PR DESCRIPTION
**Pull request by ahus1**
Closes #14734

This is currently very bold and draft attempt on building a global locking provider on top of JPA Map storage.

Previous implementations built on top Liquibase and added custom entries to the lock table which were at best surprising to those maintaining the code. This is an attempt to build the global locking like any other map store.

### CODE REVIEW

**Comment by ahus1** [comment]
@hmlnarik / @mhajas - just a short heads up: the following commit b76bad8eabed689c411617c0490c6c454435d725 now removes the extra SPI by folding the necessary bits into the global lock implementation classes. 

There is now a `MapGlobalLockProviderFactory` which is both a `GlobalLockProviderFactory` and a `AbstractMapProviderFactory<GlobalLockProvider, MapLockEntity, MapLockEntity>`. 

I think I'll be implementing the two missing fields in `MapLockEntity` and then this PR will be ready for a review.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapGlobalLockProvider.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapGlobalLockProvider.java]: This has now been fixed.

File: model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java
Comment (ahus1) [reply model/build-processor/src/main/java/org/keycloak/models/map/processor/GenerateEntityImplementationsProcessor.java]: This has now been fixed.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java:233]: This has now been fixed.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockProvider.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapLockProvider.java]: This is fixed in the next commit.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockProvider.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapLockProvider.java]: The code has changed to handle this and other cases in the current commit. Please re-review.

**Comment by ahus1** [comment]
@hmlnarik / @mhajas, this is now ready for review. Previous comments have been addressed, and I've reviewed all JavaDoc for new classes that I added. 

The last commit was just squashing the previous commits and cleaning up unused imports. As the the previous build was green, I hope this one will be green as well.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java:128]: nitpick:
```suggestion
    @Override
    public String getName() {
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java:168]:
```suggestion
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof JpaLockEntity)) return false;
        return Objects.equals(getId(), ((JpaLockEntity) obj).getId());
    }
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java:30]:
```suggestion
import static org.keycloak.models.map.lock.MapLockEntity.SearchableFields;
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/delegate/JpaLockDelegateProvider.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/delegate/JpaLockDelegateProvider.java:2]: nitpick:
```suggestion
 * Copyright 2023 Red Hat, Inc. and/or its affiliates
```

File: model/map-jpa/src/main/resources/META-INF/jpa-locks-changelog.xml
Comment (vramik) [inline model/map-jpa/src/main/resources/META-INF/jpa-locks-changelog.xml:3]: nitpick
```suggestion
Copyright 2023 Red Hat, Inc. and/or its affiliates
```

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java:2]: nitpick:
```suggestion
 * Copyright 2023 Red Hat, Inc. and/or its affiliates
```

File: model/map/src/main/resources/META-INF/services/org.keycloak.models.locking.GlobalLockProviderFactory
Comment (vramik) [inline model/map/src/main/resources/META-INF/services/org.keycloak.models.locking.GlobalLockProviderFactory:2]: nitpick
```suggestion
# Copyright 2023 Red Hat, Inc. and/or its affiliates
```

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you @ahus1 for the PR! I've several comments, mostly nitpicks, please see inline.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java:35]: Thanks, will be fixed in next commit.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java:40]: Thanks, will be fixed in next commit - this is probably due to auto-ordering which went wrong.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/entity/JpaLockEntity.java:168]: Oops, well spotted, will be fixed in next commit

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java:30]: fixed in next commit.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/delegate/JpaLockDelegateProvider.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/delegate/JpaLockDelegateProvider.java:2]: Fixed in next commit

File: model/map-jpa/src/main/resources/META-INF/locks/jpa-locks-changelog-1.xml
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/locks/jpa-locks-changelog-1.xml:45]: fixed in next commit

File: model/map-jpa/src/main/resources/META-INF/locks/jpa-locks-changelog-1.xml
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/locks/jpa-locks-changelog-1.xml:3]: fixed in next commit

File: testsuite/utils/src/main/resources/META-INF/keycloak-server.json
Comment (ahus1) [reply testsuite/utils/src/main/resources/META-INF/keycloak-server.json:41]: Good point, the next commit will add it also to the pom.xml for the integration tests so that the integration tests when running with JPA use it as well.

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorage.java
Comment (ahus1) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorage.java:123]: Good point, adding it in the next commit.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java:2]: Fixed in next commit.

File: model/map/src/main/resources/META-INF/services/org.keycloak.models.locking.GlobalLockProviderFactory
Comment (ahus1) [reply model/map/src/main/resources/META-INF/services/org.keycloak.models.locking.GlobalLockProviderFactory:2]: Fixed in next commit.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java:62]: @ahus1 Having `name` and `id` for this entity seems redundant as there are stores that don't need UUID as `id`. Can we have only `id` here and generate UUID only for JPA? This way we could also get rid of one `SearchableField`.

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @ahus1, the code looks great, I will try to find some time and play with this for HotRod. Could you please have a look at my comments?

File: model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java:201]: @mhajas - I still think it is a good way to create a meaningful unique name for the instance name as it will not only be instances fighting for locks with other instances, but also threads fighting for locks with other threads. If someone would do debgging on the instance identified with the host name, they could create a thread dump and point to the thread with the lock by using this information. 

Therefore I'd like to keep the thread name.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java:62]: The name is the name of the lock, which is defined by the caller. It can't be passed down as the ID to the store implementation, as the `#MapKeycloakTransaction#create()` is documented as follows: 

>  The ID of the entity passed in the parameter might change to a different value in the returned value if the underlying storage decided this was necessary.

As the name of the lock must be stored as-is in the store to effectively create a lock, the separate name field is required.

**Comment by ahus1** [comment]
@vramik - the tests are green again, and all your suggested changes are should now be in the branch.

**Review by vramik — COMMENTED** [review]
Thanks @ahus1. LGTM! All my comments could be resolved. I've just noticed indentation changes in model tests in `JpaMapStorage` and `JpaMapStorageCockroachdb` 

The result looks like (follow the link to see it): https://github.com/keycloak/keycloak/blob/759d07ed36e06f521e04517c22fc49e33216322d/testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorage.java#L106-L122

see `.config` part. Do you think it's worth to re-align it?

**Review by vramik — DISMISSED** [review]
Thank you @ahus1!

**Comment by ahus1** [comment]
@mhajas - the webauthn tests were unstable. If everything else is fine with this PR, please rerun the failed actions. Thanks!

File: model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java:83]: Should we use `Retry.executeWithBackoff`? Similarly as [here](https://github.com/keycloak/keycloak/blob/main/model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/locking/HotRodLocksUtils.java#L69)?

File: model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapLockEntity.java:62]: You are right, I missed it. While I don't like the fact we have two searchable fields, where one of them will be probably never used, it seems there is not any way to use only `id` or only `name`. The only thing we can do is merge those on the storage layer and return the same for both `id` and `name`.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java
Comment (mhajas) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java:46]: It is ok if this stays as is, but out of curiosity:
If I understood correctly, in JPA criteria we want to cover only fields+operators that are currently used in the codebase. Should we rather throw an exception if the ID searchable field is used?

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/lock/JpaLockModelCriteriaBuilder.java]: I agree this can be removed. Pushed a new commit which removes this searchable field even from `MapLockEntity`.

File: model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java
Comment (ahus1) [reply model/map/src/main/java/org/keycloak/models/map/lock/MapGlobalLockProvider.java]: Thanks for this hint, I've update the code in the most recent commit. I'm now waiting for the build to complete.

**Comment by ahus1** [comment]
@mhajas - this is now green for all bits except WebAuthn which is unstable. Please give it another review round when you have the time. Thanks!

**Review by mhajas — APPROVED** [review]
Thanks @ahus1, LGTM. @vramik Could you please reapprove if you agree with the latest changes?

**Review by vramik — APPROVED** [review]
lgtm

**Comment by ahus1** [comment]
@mhajas @vmuzikar - the build is now green after the change in main, so it can be merged when you think it is ready.

**Pull request merged**
